Chap_API_Key_Value_Mgmt.tex

```diff
@@ -611,6 +611,8 @@ \subsection{\code{PMIx_Fence_nb}}
 
 \adviceimplstart
 We recommend that implementation of the \refattr{PMIX_TIMEOUT} attribute be left to the host environment due to race condition considerations between completion of the operation versus internal timeout in the \ac{PMIx} server library. Implementers that choose to support \refattr{PMIX_TIMEOUT} directly in the \ac{PMIx} server library must take care to resolve the race condition and should avoid passing \refattr{PMIX_TIMEOUT} to the host environment so that multiple competing timeouts are not created.
+
+Note that \ac{PMIx} libraries may choose to implement an optimization for the case where only the calling process is involved in the fence operation by immediately returning \refconst{PMIX_OPERATION_SUCCEEDED} from the client's call in lieu of passing the fence operation to a \ac{PMIx} server.
```

Comment:
Is this also true if all of the fence participants are on the same node? Just curious since there is no data to exchange between servers the upcall into the RM is not necessary, but the server side PMIx library needs to coordinate the local distribution of data.

Reply:
I don't believe that should be allowed as the "fence" isn't just about passing data - it is an actual synchronizing barrier. This optimization only works for the case of a single process because there is no inter-process barrier to execute.

Reply:
Do you think we need to add wording to make that clear?

Reply:
I think it's ok as it is.

I was just thinking about the "all participants on the same node" case from the RM's perspective. In which case the synchronization (and possible data exchange) could happen all in the PMIx library without the upcall into the RM. The PMIx library might execute code on its internal server side, but there might not be a need to trigger the upcall to the RM (unless the RM wanted it for tracking purposes). Just something to think about I guess.

Reply:
Hmmm...I think maybe I do need to add a few words here. The optimization being discussed involved resolving the fence in the PMIx client library, thus avoiding any communication to the PMIx server library. This can only be done for the 1-proc case due to the barrier requirement.

If a PMIx server notes that all participants are on the local node, then it has no need to relay the fence operation to its host environment as there is no cross-node operation required. This is a separate optimization that I'll add to the doc. IIRC, we already implement that one (but not the 1-proc case).